### PR TITLE
 First version of the Single Body Fluid Dynamics FMI plugin

### PR DESCRIPTION
**Pull request by traversaro**
This plugin is useful for co-simulating the interaction between a (single) link
in the model and a surrounding medium. It can be used for simple lumped parameters
models, or eventually also for co-simulation with CFD solvers, but the latter case
has not been tested and it would probably be extremely slow.

For more info, see the docs: 
* https://github.com/robotology/gazebo-fmi/blob/febb3c42b38b8bbd7c734d0af13286df69a98729/plugins/single-body-fluid-dynamics/README.md
* https://github.com/robotology/gazebo-fmi/blob/febb3c42b38b8bbd7c734d0af13286df69a98729/plugins/single-body-fluid-dynamics/doc/modelica-single-body-fluid-dynamics.md

The PR https://github.com/robotology/gazebo-fmi/pull/40 needs to be merged before merging this one, so the actual difference of this PR is actually in the commit https://github.com/robotology/gazebo-fmi/pull/41/commits/febb3c42b38b8bbd7c734d0af13286df69a98729 .

### CODE REVIEW

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    ignition::math::Matrix3d world_R_link = link_R_world.Transposed();
+    ignition::math::Vector3d worldFluidForce = world_R_link*linkFluidForce;
+    ignition::math::Vector3d worldFluidMoment = world_R_link*linkFluidMoment;
+
+    link->AddForce(worldFluidForce);
+    link->AddTorque(worldFluidMoment);
+}
+
+//////////////////////////////////////////////////
+gazebo::physics::LinkPtr FMISingleBodyFluidDynamicsPlugin::FindLinkInModel(const std::string& linkName, const gazebo::physics::ModelPtr _parent)
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc]: Good  point.

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    private: gazebo::physics::LinkPtr FindLinkInModel(const std::string& linkName, const gazebo::physics::ModelPtr _parent);
+
+    /// \brief Callback on world update begin
+    private: void WorldUpdateBeginCallback(const gazebo::common::UpdateInfo & updateInfo);
+
+    /// \brief The link of which we want to simulate the fluid dynamic forces
+    private: gazebo::physics::LinkPtr link;
+
+    /// \brief FMU
+    private: FMUSingleBodyFluidDynamicsProperties m_fmu;
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh]: Yeah, actually there is a bunch of duplicate code that should properly refactored in view  of https://github.com/robotology/gazebo-fmi/issues/19 .

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    linkFluidForce[1] = m_fmu.outputVarBuffers[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicForce_y];
+    linkFluidForce[2] = m_fmu.outputVarBuffers[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicForce_z];
+    linkFluidMoment[0] = m_fmu.outputVarBuffers[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_x];
+    linkFluidMoment[1] = m_fmu.outputVarBuffers[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_y];
+    linkFluidMoment[2] = m_fmu.outputVarBuffers[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_z];
+
+    // Rotate the forces with the world orientation
+    ignition::math::Matrix3d world_R_link = link_R_world.Transposed();
+    ignition::math::Vector3d worldFluidForce = world_R_link*linkFluidForce;
+    ignition::math::Vector3d worldFluidMoment = world_R_link*linkFluidMoment;
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc]: The documentation of this Gazebo methods is unfortunately not really clear, but as far as I know and I was able to check basically the `addForce`/`addTorque` actually follow the "mixed" convention, i.e. the orientation of the world but the torque is relative to the origin of the link. 
* https://bitbucket.org/osrf/gazebo/issues/2017/improve-documentation-on-link 
* http://answers.gazebosim.org/question/16968/confused-with-addlinkforce/

**Comment by traversaro** [comment]
No problem @gabrielenava , @prashanthr05 already reviewed and I still need to address it. : )

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  }
+}
+
+//////////////////////////////////////////////////
+bool FMISingleBodyFluidDynamicsPlugin::LoadFMUs(gazebo::physics::ModelPtr _parent)
+{
+    double simulatedTimeInSeconds  = _parent->GetWorld()->SimTime().Double();
+
+    std::string instanceName = link->GetScopedName()+"_fmuSingleBodyFluidDynamics";
+    bool ok = m_fmu.fmu.load(m_fmu.fmuAbsolutePath, instanceName, simulatedTimeInSeconds);
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc]: I am not sure what you meant here. Which aspect to you think it make sense to comment?

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    /// \brief An identifier for the actuator.
+    public: std::string name;
+
+    public: std::string fmuAbsolutePath;
+
+    /// \brief Default input variable names
+    public: std::vector<std::string> m_inputVariablesDefaultNames;
+
+    /// \brief Actual input variable names, after parsing variable_names
+    public: std::vector<std::string> m_inputVariablesNames;
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh]: Exactly, that is the idea.

File: plugins/single-body-fluid-dynamics/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+For a precise definition of what the `causality` of a FMU variable is, see Section 3.2 of [the "Functional Mockup Interface 2.0: The Standard for Tool independent Exchange of Simulation Models" paper](http://lup.lub.lu.se/search/ws/files/5428900/2972293.pdf) or [Section 2.2.7 of the "Functional Mock-up Interface for
+Model Exchange and Co-Simulation" v2.0 specification](https://fmi-standard.org/docs/2.0.1-develop/#_definition_of_model_variables_modelvariables).
+
+
+## Operating principle
+The `gazebo-fmi` single body fluid dynamics plugin code is called during in the `worldUpdateBegin` event.
+The plugin reads the velocity of the body using and the velocity of the wind, and computes from this the relative velocity.
+
+The working of the plugin was inspired by the Gazebo plugin LiftDragPlugin, described in the [Gazebo's Aerodynamics tutorial](http://gazebosim.org/tutorials?tut=aerodynamics&cat=plugins).
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: I actually changed it to `LiftDragPlugin`, I think it fits well here.

File: plugins/single-body-fluid-dynamics/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  <plugin name="fmi_aero_plugin" filename="libFMISingleBodyFluidDynamicsPlugin.so">
+    <single_body_fluid_fynamics>
+      <name>drag_0</name> 
+      <link>link0</link>
+      <fmu>drag.fmu</fmu>
+    </single_body_fluid_fynamics>
+  </plugin>
+</model>
+~~~
+The plugin filename is `libFMISingleBodyFkuidDynamicsPlugin.so` .
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+Example configuration: 
+~~~
+<model>
+  ...
+  <plugin name="fmi_aero_plugin" filename="libFMISingleBodyFluidDynamicsPlugin.so">
+    <single_body_fluid_fynamics>
+      <name>drag_0</name> 
+      <link>link0</link>
+      <fmu>drag.fmu</fmu>
+    </single_body_fluid_fynamics>
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+of the relative velocity between the body and the fluid. 
+As it is possible to understand from the name, this plugin does not permit to simulate the mutual interaction of two or more bodies close to  each other with the surronding medium.
+
+## Use the plugin 
+Example configuration: 
+~~~
+<model>
+  ...
+  <plugin name="fmi_aero_plugin" filename="libFMISingleBodyFluidDynamicsPlugin.so">
+    <single_body_fluid_fynamics>
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+Documentation of the parameters of the `<single_body_fluid_dynamics>` tag. All the parameters are required
+
+| Parameter name | Type    | Description                 | Notes |
+|:--------------:|:-------:|:--------------------------: |:-----:|
+| name           | string  | Name of the actuator, used for printing debug and error messages. |  |
+| link          | string  | Name of the link. | The total list of joints contained in the model is scanned and the first joint that **link** with this  name string is found. This is done to easily support nested models. Alternatively you can specify directly the **scoped link name** as well.   |
+| fmu            | string  | Filename of the FMU plugin to use for actuator co-simulation. | This name is passed to the [`gazebo::common::SystemPaths::FindFile`](http://osrf-distributions.s3.amazonaws.com/gazebo/api/9.0.0/classgazebo_1_1common_1_1SystemPaths.html#a9e03f07eac9f89d8c4c14af5660fa938) method to find the absolute location of the FMU file. Adding the directory containing the FMUs to the [`GAZEBO_RESOURCE_PATH`](http://gazebosim.org/tutorials?tut=components) should be sufficient to make it visible to the plugin. |
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Basically one  of the reasons is that in the Gazebo API there is no `FindFileInPluginPaths` method.

File: plugins/single-body-fluid-dynamics/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+| variable_names            | string  | Filename of the FMU plugin to use for actuator co-simulation. | This name is passed to the [`gazebo::common::SystemPaths::FindFile`](http://osrf-distributions.s3.amazonaws.com/gazebo/api/9.0.0/classgazebo_1_1common_1_1SystemPaths.html#a9e03f07eac9f89d8c4c14af5660fa938) method to find the absolute location of the FMU file. Adding the directory containing the FMUs to the [`GAZEBO_RESOURCE_PATH`](http://gazebosim.org/tutorials?tut=components) should be sufficient to make it visible to the plugin. |
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+    public: FMUCoSimulation fmu;
+    public: std::vector<fmi2_value_reference_t> inputVarReferences;
+    public: std::vector<fmi2_value_reference_t> outputVarReferences;
+    public: std::vector<double> inputVarBuffers;
+    public: std::vector<double> outputVarBuffers;
+
+};
+
+/// \brief Plugin for simulating a torque-speed curve for actuators.
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+
+Variables:
+| Parameter name | Type    | Description                 | Notes |
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+|:--------------:|:-------:|:--------------------------: |:-----:|
+| `relativeVelocity_x` | X component of the relative velocity between the body and the medium, expressed in the body frame. | This variable should be present in the FMU with causality `input`, and is expressed in m/s. |
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+| `fluidDynamicForce_x`   | X component of the aerodynamic force exerted by the medium on body, expressed in the body frame. | This variable should be present in the FMU with causality `output`,  and  is expressed in N. |
+| `fluidDynamicForce_y`   | Y component of the aerodynamic force exerted by the medium on body, expressed in the body frame. | This variable should be present in the FMU with causality `output`, and  is expressed in N. |
+| `fluidDynamicForce_z`   | Z component of the aerodynamic force exerted by the medium on body, expressed in the body frame. | This variable should be present in the FMU with causality `output`,  and  is expressed in N. |
+| `fluidDynamicMoment_x`   | X component of the aerodynamic moment exerted by the medium on body, expressed in the body orientation and w.r.t. to the boy origin. | This variable should be present in the FMU with causality `output`, and  is expressed in Nm. |
+| `fluidDynamicMoment_y`   | Y component ofthe aerodynamic moment exerted by the medium on body, expressed in the  body orientation and w.r.t. to the boy origin. | This variable should be present in the FMU with causality `output`, and  is expressed in Nm. |
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+## Operating principle
+The `gazebo-fmi` single body fluid dynamics plugin code is called during in the `worldUpdateBegin` event.
+The plugin reads the velocity of the body using and the velocity of the wind, and computes from this the relative velocity.
+
+The working of the plugin was inspired by the Gazebo plugin LiftDragPlugin, described in the [Gazebo's Aerodynamics tutorial](http://gazebosim.org/tutorials?tut=aerodynamics&cat=plugins).
+
+## Modelica examples
+The gazebo-fmi plugins can be used with any standard-compliant FMU.
+For testing purpouses we generate fluid dynamics FMUs with [Modelica](https://www.modelica.org/), that is
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+| `fluidDynamicMoment_y`   | Y component ofthe aerodynamic moment exerted by the medium on body, expressed in the  body orientation and w.r.t. to the boy origin. | This variable should be present in the FMU with causality `output`, and  is expressed in Nm. |
+| `fluidDynamicMoment_z`   | Z component of the aerodynamic moment exerted by the medium on body, expressed in the  body orientation and w.r.t. to the boy origin.| This variable should be present in the FMU with causality `output`, and  is expressed in Nm. |
+
+For a precise definition of what the `causality` of a FMU variable is, see Section 3.2 of [the "Functional Mockup Interface 2.0: The Standard for Tool independent Exchange of Simulation Models" paper](http://lup.lub.lu.se/search/ws/files/5428900/2972293.pdf) or [Section 2.2.7 of the "Functional Mock-up Interface for
+Model Exchange and Co-Simulation" v2.0 specification](https://fmi-standard.org/docs/2.0.1-develop/#_definition_of_model_variables_modelvariables).
+
+
+## Operating principle
+The `gazebo-fmi` single body fluid dynamics plugin code is called during in the `worldUpdateBegin` event.
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/README.md [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+The plugin reads the velocity of the body using and the velocity of the wind, and computes from this the relative velocity.
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/README.md]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+#include <string>
+#include <gazebo/common/Events.hh>
+#include <gazebo/physics/physics.hh>
+#include <gazebo/gazebo.hh>
+
+#include <gazebo_fmi/FMUCoSimulation.hh>
+
+namespace gazebo_fmi
+{
+/// \brief Properties for a model of a actuator
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+class FMUSingleBodyFluidDynamicsProperties
+{
+    /// \brief An identifier for the actuator.
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.hh]: Fixed in https://github.com/robotology/gazebo-fmi/pull/41/commits/f05bcff1eb2aed1c99fe9b3943a41b5a9537978d .

File: plugins/single-body-fluid-dynamics/doc/modelica-single-body-fluid-dynamics.md [hunk 21]
```diff
@@ -0,0 +1,48 @@
+# Use of the gazebo-fmi Single Body Fluid Dynamics plugin with Modelica
+
+**Note: this document discusses the use Modelica with the gazebo-fmi Single Body Fluid Dynamics plugin, using [OpenModelica](https://openmodelica.org/). The content should apply also to other Modelica implementation that support the FMI 2.0 standard, but their use was not tested.**
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/doc/modelica-single-body-fluid-dynamics.md]: Done in https://github.com/robotology/gazebo-fmi/pull/41/commits/d433ff5c6275a558543c59ca25847289f4b3ba2b

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+    m_fmu.m_outputVariablesDefaultNames[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicForce_y] = "fluidDynamicForce_y";
+    m_fmu.m_outputVariablesDefaultNames[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicForce_z] = "fluidDynamicForce_z";
+    m_fmu.m_outputVariablesDefaultNames[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_x] = "fluidDynamicMoment_x";
+    m_fmu.m_outputVariablesDefaultNames[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_y] = "fluidDynamicMoment_y";
+    m_fmu.m_outputVariablesDefaultNames[FMISingleBodyFluidDynamicsPluginNS::fluidDynamicMoment_z] = "fluidDynamicMoment_z";
+
+    // Bullet is not supported by this plugin, due to https://bitbucket.org/osrf/gazebo/issues/1476/implement-addxxxforce-for-bullet
+    if (gazebo::physics::get_world()->Physics()->GetType() == "bullet")
+    {
+        gzerr << "FMISingleBodyFluidDynamicsPlugin: the bullet physics engine is not supported, as in Gazebo there is no way to excert an external force on a link." << std::endl;
```
[the remark this answers is not in the document]
Comment (traversaro) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc]: Done in https://github.com/robotology/gazebo-fmi/pull/41/commits/594a2c9b6d240ce9f5f1f72bcd65f75937490661 .

**Comment by traversaro** [comment]
I should have addressed all the reviews.

File: plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (prashanthr05) [reply plugins/single-body-fluid-dynamics/FMISingleBodyFluidDynamicsPlugin.cc]: if I remember right, the FMU model is being installed at a certain location, using the FMILibrary right? I think what I had in mind was that we are trying to access the pre-compiled FMU in order to load the FMU instance.

**Comment by traversaro** [comment]
@prashanthr05 Can you check if I address your reviews? Thanks!

**Comment by prashanthr05** [comment]
@traversaro I am sorry for the delay. I was thinking I had approved this request. 

However, looks like Travis is failing?

**Comment by traversaro** [comment]
> However, looks like Travis is failing for gazebo-7 ? Are we supporting Gazebo 7?

Discussion on this in https://github.com/robotology/gazebo-fmi/pull/44 .

**Comment by traversaro** [comment]
Rebased on the top of https://github.com/robotology/gazebo-fmi/pull/29, this means that this needs to be merged after https://github.com/robotology/gazebo-fmi/pull/29 .

**Pull request merged**